lib/Model/Type.cpp

```diff
@@ -648,28 +649,91 @@ QualifiedType::size(VerifyHelper &VH) const {
   rc_return rc_recur UnqualifiedType.get()->size(VH);
 }
 
-inline RecursiveCoroutine<bool>
-isPrimitive(const model::QualifiedType &QT,
-            model::PrimitiveTypeKind::Values V) {
-  auto IsConstQualifier = [](const Qualifier &Q) {
-    return Q.Kind == model::QualifierKind::Const;
-  };
+template<bool TraverseTypedefs = true>
+static RecursiveCoroutine<bool>
+isEventuallyArrayImpl(const model::QualifiedType &QT) {
+  const auto &IsNonConst = std::not_fn(model::Qualifier::isConst);
+  for (const auto &Q : llvm::make_filter_range(QT.Qualifiers, IsNonConst)) {
+
+    // If we find an array first, it's definitely an array, otherwise we
+    // found a pointer first, so it's definitely not an array
+    if (Qualifier::isArray(Q))
+      rc_return true;
+
+    rc_return false;
+  }
+
+  if constexpr (TraverseTypedefs)
+    if (auto *TD = dyn_cast<model::TypedefType>(QT.UnqualifiedType.get()))
+      rc_return rc_recur isEventuallyArrayImpl(TD->UnderlyingType);
+
+  // If there are no non-const qualifiers, it's not an array
+  rc_return false;
+}
+
+bool QualifiedType::isArray() const {
+  return isEventuallyArrayImpl<false>(*this);
+}
+
+bool QualifiedType::isEventuallyArray() const {
+  return isEventuallyArrayImpl<true>(*this);
+}
+
+template<bool TraverseTypedefs = true>
+static RecursiveCoroutine<bool>
+isEventuallyPointerImpl(const model::QualifiedType &QT) {
+  const auto &IsNonConst = std::not_fn(Qualifier::isConst);
```

Comment:
Wouldn't `NotIsConst` be more appropriate?

Reply:
Will do

Reply:
Done!